README.md
# stactools-alos-dem

- Name: alos-dem
- Package: `stactools.alos_dem`
- PyPI: https://pypi.org/project/stactools-alos-dem/
- Owner: @gadomski
- Dataset homepages:
  - https://www.eorc.jaxa.jp/ALOS/en/aw3d30/index.htm
  - https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.112016.4326.2
- STAC extensions used:
  - [proj](https://github.com/stac-extensions/projection/)
- Extra fields: none

This data set is a global digital surface model (DSM) with horizontal resolution of approximately 30 meters (basically 1 arcsecond) by the Panchromatic Remote-sensing Instrument for Stereo Mapping (PRISM), which was an optical sensor on board the Advanced Land Observing Satellite "ALOS".

(from https://www.eorc.jaxa.jp/ALOS/en/aw3d30/index.htm)

[![CI](https://github.com/stactools-packages/alos-dem/actions/workflows/continuous-integration.yml/badge.svg)](https://github.com/stactools-packages/alos-dem/actions/workflows/continuous-integration.yml)
[![Binder](https://mybinder.org/badge_logo.svg)](https://mybinder.org/v2/gh/stactools-packages/alos-gap/main?filepath=docs/installation_and_basic_usage.ipynb)
## Examples

### STAC objects

- [Item](examples/ALPSMLC30_N041W106_DSM.json)

### Command-line usage

Description of the command line functions

```bash
$ stac alos-dem create-item ALPSMLC30_N041W106_DSM.tif ALPSMLC30_N041W106_DSM.json
```

Use `stac package --help` to see all subcommands and options.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/). This project attempts to match the major and minor versions of [stactools](https://github.com/stac-utils/stactools) and increments the patch number as needed.

## [v0.1.2] - 2021-08-20

### Fixed

- Moved s3 to be a dev dependency only ([#6](https://github.com/stactools-packages/alos-dem/pull/6)).
  Includes a manual pin of the **aiobotocore** version to work around https://github.com/dask/s3fs/pull/510, which should be reverted once that PR is released.

## [v0.1.1] - 2021-08-02

### Added

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Aligned platform and instruments fields with best practices.

## [v0.1.0] - 2021-07-29

The initial release of this package.
### Added

- Everything.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.


tests/__init__.py
from stactools.testing import TestData

test_data = TestData(
    __file__, {
        "ALPSMLC30_N041W106_DSM.tif": {
            "url":
            "s3://raster/AW3D30/AW3D30_global/ALPSMLC30_N041W106_DSM.tif",
            "s3": {
                "anon": True,
                "client_kwargs": {
                    "endpoint_url": "https://opentopography.s3.sdsc.edu",
                }
            }
        }
    })


tests/test_stac.py
import datetime
from unittest import TestCase

from pystac import Provider, MediaType
from pystac.extensions.projection import ProjectionExtension

from stactools.alos_dem import stac

from tests import test_data


class StacTest(TestCase):
    def setUp(self):
        self.path = test_data.get_external_data("ALPSMLC30_N041W106_DSM.tif")

    def test_create_item(self):
        item = stac.create_item(self.path)
        self.assertEqual(item.id, "ALPSMLC30_N041W106_DSM")
        self.assertIsNotNone(item.geometry)
        self.assertEqual(list(item.bbox), [-106.0, 41.0, -105.0, 42.0])
        self.assertEqual(
            item.datetime,
            datetime.datetime(2016, 12, 7, tzinfo=datetime.timezone.utc))

        common_metadata = item.common_metadata
        self.assertEqual(common_metadata.platform, "alos")
        self.assertEqual(common_metadata.instruments, ["prism"])
        self.assertEqual(common_metadata.gsd, 30)
        self.assertEqual(common_metadata.providers, [
            Provider("Japan Aerospace Exploration Agency",
                     roles=["producer", "processor", "licensor"],
                     url="https://www.eorc.jaxa.jp/ALOS/en/aw3d30/index.htm"),
            Provider("OpenTopography",
                     roles=["host"],
                     url=("https://portal.opentopography.org/"
                          "raster?opentopoID=OTALOS.112016.4326.2")),
        ])
        self.assertEqual(common_metadata.license, "proprietary")

        projection = ProjectionExtension.ext(item)
        self.assertEqual(projection.epsg, 4326)
        self.assertEqual(projection.shape, (3600, 3600))
        self.assertEqual(list(projection.transform), [
            0.0002777777777777778, 0.0, -106.0, 0.0, -0.0002777777777777778,
            42.0
        ])

        handbook = item.get_single_link("handbook")
        self.assertIsNotNone(handbook)
        self.assertEqual(handbook.title, "ALOS User handbook")
        self.assertEqual(handbook.rel, "handbook")
        self.assertEqual(
            handbook.href,
            "https://www.eorc.jaxa.jp/ALOS/en/doc/alos_userhb_en.pdf")
        self.assertEqual(handbook.media_type, "application/pdf")

        data = item.assets["data"]
        self.assertEqual(data.href, self.path)
        self.assertEqual(data.title, "N041W106")
        self.assertIsNone(data.description)
        self.assertEqual(data.media_type, MediaType.COG)
        self.assertEqual(data.roles, ["data"])

        item.validate()

    def test_create_item_with_read_href_modifier(self):
        done = False

        def do_it(href):
            nonlocal done
            done = True
            return href

        _ = stac.create_item(self.path, read_href_modifier=do_it)
        self.assertTrue(done, "Didn't do it")


tests/test_module.py
import unittest

import stactools.alos_dem


class TestModule(unittest.TestCase):
    def test_version(self):
        self.assertIsNotNone(stactools.alos_dem.__version__)


src/stactools/alos_dem/stac.py
import os.path
from typing import Optional
from pystac.asset import Asset
from pystac.extensions.projection import ProjectionExtension
from pystac.media_type import MediaType

import rasterio
from shapely.geometry import mapping, box
from pystac import Item

from stactools.core.io import ReadHrefModifier
from stactools.alos_dem.constants import (ALOS_DEM_LINKS, ALOS_DEM_PROVIDERS,
                                          OPENTOPOGRAPHY_DATETIME,
                                          ALOS_DEM_PLATFORM,
                                          ALOS_DEM_INSTRUMENTS, ALOS_DEM_GSD,
                                          ALOS_DEM_EPSG)


def create_item(href: str,
                read_href_modifier: Optional[ReadHrefModifier] = None) -> Item:
    """Creates a STAC Item from a single tile of ALOS DEM data."""
    if read_href_modifier:
        modified_href = read_href_modifier(href)
    else:
        modified_href = href
    with rasterio.open(modified_href) as dataset:
        if dataset.crs.to_epsg() != 4326:
            raise ValueError(
                f"Dataset {href} is not EPSG:4326, which is required for ALOS DEM data"
            )
        bbox = list(dataset.bounds)
        geometry = mapping(box(*bbox))
        transform = dataset.transform
        shape = dataset.shape
    item = Item(id=os.path.splitext(os.path.basename(href))[0],
                geometry=geometry,
                bbox=bbox,
                datetime=OPENTOPOGRAPHY_DATETIME,
                properties={},
                stac_extensions={})

    item.add_links(ALOS_DEM_LINKS)
    item.common_metadata.platform = ALOS_DEM_PLATFORM
    item.common_metadata.instruments = ALOS_DEM_INSTRUMENTS
    item.common_metadata.gsd = ALOS_DEM_GSD
    item.common_metadata.providers = ALOS_DEM_PROVIDERS
    item.common_metadata.license = "proprietary"

    parts = os.path.basename(href).split("_")
    if len(parts) != 3:
        raise ValueError(
            f"Unexpected file name, expected two underscores in name: {os.path.basename(href)}"
        )
    title = parts[1]
    item.add_asset(
        "data",
        Asset(href=href,
              title=title,
              description=None,
              media_type=MediaType.COG,
              roles=["data"]))

    projection = ProjectionExtension.ext(item, add_if_missing=True)
    projection.epsg = ALOS_DEM_EPSG
    projection.transform = transform[0:6]
    projection.shape = shape

    return item


src/stactools/alos_dem/constants.py
import datetime

from pystac import Provider, Link

ALOS_DEM_PLATFORM = "alos"
ALOS_DEM_INSTRUMENTS = ["prism"]
ALOS_DEM_GSD = 30  # meters
ALOS_DEM_EPSG = 4326
ALOS_DEM_PROVIDERS = [
    Provider("Japan Aerospace Exploration Agency",
             roles=["producer", "processor", "licensor"],
             url="https://www.eorc.jaxa.jp/ALOS/en/aw3d30/index.htm"),
    Provider("OpenTopography",
             roles=["host"],
             url=("https://portal.opentopography.org/"
                  "raster?opentopoID=OTALOS.112016.4326.2")),
]
ALOS_DEM_LINKS = [
    Link("handbook",
         "https://www.eorc.jaxa.jp/ALOS/en/doc/alos_userhb_en.pdf",
         "application/pdf",
         "ALOS User handbook",
         extra_fields={"description": "Also includes data usage information"})
]

# This stactools package was created to interact with the ALOS DEM data hosted
# on OpenTopography. As of this writing, the data were last updated at this
# time. This information was taken from
# https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.112016.4326.2.
OPENTOPOGRAPHY_DATETIME = datetime.datetime(2016,
                                            12,
                                            7,
                                            tzinfo=datetime.timezone.utc)
